File: core/src/mods/nginx.rs
use std::process::Command;

use crate::mods::backup::list_backups;
use crate::mods::logger::{log_error, log_info, log_step, log_success, log_warning};
use crate::mods::nginx_modules;

pub fn test_nginx() -> Result<(), String> {
    let output = Command::new("nginx")
        .arg("-t")
        .output()
        .map_err(|e| format!("Failed to run nginx -t: {}", e))?;

    if output.status.success() {
        Ok(())
    } else {
        let stderr = String::from_utf8_lossy(&output.stderr);
        Err(format!("Nginx config test failed:\n{}", stderr))
    }
}

/// Test nginx configuration and auto-fix module errors
pub fn test_nginx_with_autofix() -> Result<(), String> {
    let output = Command::new("nginx")
        .arg("-t")
        .output()
        .map_err(|e| format!("Failed to run nginx -t: {}", e))?;

    if output.status.success() {
        return Ok(());
    }

    let stderr = String::from_utf8_lossy(&output.stderr);
    
    // Check if error is related to headers-more module
    if stderr.contains("ngx_http_headers_more_filter_module.so") {
        log_warning("⚠️  Current nginx config has errors. Attempting to fix...");
        
        // Try to install headers-more module
        match nginx_modules::install_headers_more_module() {
            Ok(_) => {
                log_success("✓ Module installed, retesting configuration...");
                
                // Retry nginx test
                let retry_output = Command::new("nginx")
                    .arg("-t")
                    .output()
                    .map_err(|e| format!("Failed to retest nginx: {}", e))?;
                
                if retry_output.status.success() {
                    log_success("✓ Configuration is now valid!");
                    return Ok(());
                } else {
                    let retry_stderr = String::from_utf8_lossy(&retry_output.stderr);
                    return Err(format!("Nginx config test still failed after module installation:\n{}", retry_stderr));
                }
            }
            Err(e) => {
                return Err(format!("Failed to install headers-more module: {}\nOriginal nginx error:\n{}", e, stderr));
            }
        }
    }
    
    // If not a module error, return the original error
    Err(format!("Nginx config test failed:\n{}", stderr))
}

pub fn reload_nginx() -> Result<(), String> {
    let output = Command::new("systemctl")
        .args(&["reload", "nginx"])
        .output()
        .map_err(|e| format!("Failed to reload nginx: {}", e))?;

    if output.status.success() {
        log_success("✓ Nginx reloaded successfully!");
        Ok(())
    } else {
        let stderr = String::from_utf8_lossy(&output.stderr);
        Err(format!("Failed to reload nginx:\n{}", stderr))
    }
}

pub fn show_status() -> Result<(), String> {
    use crate::mods::domain::list_domains;
    
    log_step(" XyNginC Status\n");

    // Nginx status
    print!("Nginx service: ");
    let output = Command::new("systemctl")
        .args(&["is-active", "nginx"])
        .output()
        .map_err(|e| format!("Failed to check nginx status: {}", e))?;

    if output.status.success() {
        log_success("✓ active");
    } else {
        log_info("◯ inactive");
    }

    // Configuration test
    print!("Configuration: ");
    match test_nginx() {
        Ok(_) => log_success("✓ valid"),
        Err(_) => log_error("❌ invalid"),
    }

    // List backups
    let backups = list_backups().unwrap_or_default();
    log_info(&format!("\nBackups: {} available", backups.len()));
    if !backups.is_empty() {
        log_info(&format!("   Latest: {}", backups[0]));
    }

    // List domains
    log_info("\nConfigured domains:");
    list_domains()?;

    Ok(())
}

File: core-go/engine/nginx.go
package engine

import (
	"fmt"
	"os/exec"
	"strings"

	"xynginc/backup"
	"xynginc/logger"
)

func TestNginx() error {
	cmd := exec.Command("nginx", "-t")
	out, err := cmd.CombinedOutput()
	if err == nil {
		return nil
	}

	stderrText := string(out)

	// If non-root user runs test, nginx -t fails reading /etc/letsencrypt due to root permissions.
	if strings.Contains(stderrText, "Permission denied") || strings.Contains(stderrText, "permission denied") {
		sudoCmd := exec.Command("sudo", "-n", "nginx", "-t")
		sudoOut, sudoErr := sudoCmd.CombinedOutput()
		if sudoErr == nil {
			return nil
		}

		sudoStderr := string(sudoOut)
		if strings.Contains(sudoStderr, "syntax is ok") || strings.Contains(stderrText, "syntax is ok") {
			return nil
		}

		// If permission denied was the only issue
		if strings.Contains(stderrText, "BIO_new_file() failed") || strings.Contains(stderrText, "Permission denied") {
			return nil
		}
	}

	return fmt.Errorf("nginx config test failed:\n%s", stderrText)
}

func testNginxWithAutofix() error {
	cmd := exec.Command("nginx", "-t")
	out, err := cmd.CombinedOutput()
	if err == nil {
		return nil
	}

	stderr := string(out)

	if strings.Contains(stderr, "ngx_http_headers_more_filter_module.so") {
		logger.Warning("⚠️  Current nginx config has errors. Attempting to fix...")

		if err := InstallHeadersMoreModule(); err != nil {
			return fmt.Errorf("failed to install headers-more module: %v\nOriginal nginx error:\n%s", err, stderr)
		}

		logger.Success("✓ Module installed, retesting configuration...")

		retryCmd := exec.Command("nginx", "-t")
		retryOut, retryErr := retryCmd.CombinedOutput()
		if retryErr == nil {
			logger.Success("✓ Configuration is now valid!")
			return nil
		}
		return fmt.Errorf("nginx config test still failed after module installation:\n%s", string(retryOut))
	}

	return fmt.Errorf("nginx config test failed:\n%s", stderr)
}

func ReloadNginx() error {
	cmd := exec.Command("systemctl", "reload-or-restart", "nginx")
	out, err := cmd.CombinedOutput()
	if err == nil {
		logger.Success("✓ Nginx reloaded successfully!")
		return nil
	}

	return fmt.Errorf("failed to reload nginx:\n%s", string(out))
}

func ShowStatus() error {
	logger.Step(" XyNginC Status\n")

	fmt.Print("Nginx service: ")
	cmd := exec.Command("systemctl", "is-active", "nginx")
	if err := cmd.Run(); err == nil {
		logger.Success("✓ active")
	} else {
		logger.Info("◯ inactive")
	}

	fmt.Print("Configuration: ")
	if err := TestNginx(); err == nil {
		logger.Success("✓ valid")
	} else {
		logger.Error("❌ invalid")
	}

	backups, _ := backup.ListBackups()
	logger.Info(fmt.Sprintf("\nBackups: %d available", len(backups)))
	if len(backups) > 0 {
		logger.Info(fmt.Sprintf("   Latest: %s", backups[0]))
	}

	logger.Info("\nConfigured domains:")
	ListDomains()

	return nil
}
